DynamicModeler/Logic/vtkSlicerDynamicModelerRevolveTool.cxx

```diff
@@ -0,0 +1,375 @@
+#include <vtkMRMLMarkupsAngleNode.h>
+#include <vtkMRMLModelNode.h>
+#include <vtkMRMLTransformNode.h>
+
+// VTK includes
+#include <vtkCommand.h>
+#include <vtkGeneralTransform.h>
+#include <vtkIntArray.h>
+#include <vtkFloatArray.h>
+#include <vtkPointData.h>
+#include <vtkRotationalExtrusionFilter.h>
+#include <vtkPolyDataNormals.h>
+#include <vtkSmartPointer.h>
+#include <vtkStringArray.h>
+#include <vtkTransform.h>
+#include <vtkTransformPolyDataFilter.h>
+#include <vtkTriangleFilter.h>
+#include <vtkFeatureEdges.h>
+#include <vtkAppendPolyData.h>
+
+//----------------------------------------------------------------------------
+vtkToolNewMacro(vtkSlicerDynamicModelerRevolveTool);
+
+const char* REVOLVE_INPUT_MODEL_REFERENCE_ROLE = "Revolve.InputModel";
+const char* REVOLVE_INPUT_MARKUPS_REFERENCE_ROLE = "Revolve.InputMarkups";
+const char* REVOLVE_OUTPUT_MODEL_REFERENCE_ROLE = "Revolve.OutputModel";
+const char* REVOLVE_ANGLE_DEGREES = "Revolve.AngleDegrees";
+const char* REVOLVE_AXIS_AT_ORIGIN = "Revolve.AxisAtOrigin";
+const char* REVOLVE_TRANSLATE_ALONG_AXIS_DISTANCE = "Revolve.TranslateDistanceAlongAxis";
+
+//----------------------------------------------------------------------------
+vtkSlicerDynamicModelerRevolveTool::vtkSlicerDynamicModelerRevolveTool()
+{
+  /////////
+  // Inputs
+  vtkNew<vtkIntArray> inputModelEvents;
+  inputModelEvents->InsertNextTuple1(vtkCommand::ModifiedEvent);
+  inputModelEvents->InsertNextTuple1(vtkMRMLModelNode::MeshModifiedEvent);
+  inputModelEvents->InsertNextTuple1(vtkMRMLTransformableNode::TransformModifiedEvent);
+  vtkNew<vtkStringArray> inputModelClassNames;
+  inputModelClassNames->InsertNextValue("vtkMRMLModelNode");
+  NodeInfo inputModel(
+    "Model",
+    "Model to be revolved.",
+    inputModelClassNames,
+    REVOLVE_INPUT_MODEL_REFERENCE_ROLE,
+    true,
+    false,
+    inputModelEvents
+  );
+  this->InputNodeInfo.push_back(inputModel);
+
+  vtkNew<vtkIntArray> inputMarkupEvents;
+  inputMarkupEvents->InsertNextTuple1(vtkCommand::ModifiedEvent);
+  inputMarkupEvents->InsertNextTuple1(vtkMRMLMarkupsNode::PointModifiedEvent);
+  inputMarkupEvents->InsertNextTuple1(vtkMRMLTransformableNode::TransformModifiedEvent);
+  vtkNew<vtkStringArray> inputMarkupClassNames;
+  inputMarkupClassNames->InsertNextValue("vtkMRMLMarkupsFiducialNode");
+  inputMarkupClassNames->InsertNextValue("vtkMRMLMarkupsLineNode");
+  inputMarkupClassNames->InsertNextValue("vtkMRMLMarkupsPlaneNode");
+  inputMarkupClassNames->InsertNextValue("vtkMRMLMarkupsAngleNode");
+  NodeInfo inputMarkups(
+    "Markups",
+    "Markups to specify spatial revolution axis. Normal for plane and angle, superior axis for a point.",
+    inputMarkupClassNames,
+    REVOLVE_INPUT_MARKUPS_REFERENCE_ROLE,
+    /*required*/ true,
+    /*repeatable*/ false,
+    inputMarkupEvents
+  );
+  this->InputNodeInfo.push_back(inputMarkups);
+
+  /////////
+  // Outputs
+  NodeInfo outputModel(
+    "Revolved model",
```

Comment:
Most of the use cases that I can imagine would use an open or closed markup curve as input. Could you add support for that, too?

Reply:
Done